torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -34,6 +34,7 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
     return {};
 
   auto new_grad = std::move(grads[0]);
+  std::lock_guard<std::mutex> lock(grad_mutex);
   for (auto& hook : impl::hooks(variable)) {
     new_grad = (*hook)({new_grad})[0];
```

Comment:
This seems like it could cause a deadlock, if a hook triggers another invocation of backwards, which attempts to reaccumulate into the same variable.

Reply:
hmm yeah that's true. How about moving the lock after the hooks. The downside of this approach is that the hooks need to retain thread-safe fashion in multithreaded autograd.

Reply:
Yeah... Python hooks are automatically thread safe, but I guess we support C++ hooks too.

Reply:
Is that written explicitly in the doc?
Did you checked if we currently have C++ hooks that are not thread safe?